File: GanServer/src/com/xattacker/json/JsonUtility.java
package com.xattacker.json;

import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.Expose;
import com.google.gson.reflect.TypeToken;


public final class JsonUtility
{
	public static String serialize(Object obj) throws Exception
	{
		return serialize(obj, null);
	}
	
	public static String serialize(Object obj, JsonBuilderVisitor aVisitor) throws Exception
	{
//		ObjectMapper mapper = new ObjectMapper();
//		return mapper.writeValueAsString(obj);

		return createGson(aVisitor).toJson(obj);
	}
	
	public static Gson createGson(JsonBuilderVisitor aVisitor)
	{
	   GsonBuilder builder = new GsonBuilder();  
	   
	   if (aVisitor != null)
	   {
	   	aVisitor.onBuilderPrepared(builder);
	   }

		builder .setFieldNamingPolicy(FieldNamingPolicy.IDENTITY);
		
	   // ignore expose field to serialize
	   builder.setExclusionStrategies
      (
          new ExclusionStrategy()
          {
              @Override
              public boolean shouldSkipField(FieldAttributes f)
              {
                  return f.getAnnotation(Expose.class) != null;
              }

              @Override
              public boolean shouldSkipClass(Class<?> clazz)
              {
                  return false;
              }
          }
      );
	   
	   return builder.create();   
	}
	
	public static Gson createGson()
	{
		return createGson(null);
	}
	
	public static <T> T fromJson(String aJson, TypeToken<T> aToken, JsonBuilderVisitor aVisitor)
	{
		Gson gson = createGson(aVisitor);
		return gson.fromJson(aJson, aToken.getType());
	}
	
	public static <T> T fromJson(String aJson, TypeToken<T> aToken)
	{
		return fromJson(aJson, aToken, null);
	}
	
	public static <T> T fromJson(String aJson, Class<T> aType, JsonBuilderVisitor aVisitor)
	{
		Gson gson = createGson(aVisitor);
		return gson.fromJson(aJson, aType);
	}
	
	public static <T> T fromJson(String aJson, Class<T> aType)
	{
		return fromJson(aJson, aType, null);
	}
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/json/JsonUtility.kt
package com.xattacker.json

import android.content.Context
import com.google.gson.ExclusionStrategy
import com.google.gson.FieldAttributes
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.annotations.Expose
import com.google.gson.reflect.TypeToken
import java.io.ByteArrayOutputStream
import java.io.InputStream

object JsonUtility
{
    @JvmOverloads
    fun createGson(visitor: ((builder: GsonBuilder) -> Unit)? = null): Gson
    {
        val builder = GsonBuilder()
        visitor?.invoke(builder)

        // ignore expose field to serialize
        builder.setExclusionStrategies(
            object : ExclusionStrategy
            {
                override fun shouldSkipField(f: FieldAttributes): Boolean
                {
                    return f.getAnnotation<Expose>(Expose::class.java) != null
                }

                override fun shouldSkipClass(clazz: Class<*>): Boolean
                {
                    return false
                }
            }
        )

        return builder.create()
    }

    @JvmOverloads
    fun <T> fromJson(json: String, token: TypeToken<T>, visitor:((builder: GsonBuilder) -> Unit)? = null): T?
    {
        val gson = createGson(visitor)
        return gson.fromJson<T>(json, token.type)
    }

    @JvmOverloads
    fun <T> fromJson(json: String, type: Class<T>, visitor: ((builder: GsonBuilder) -> Unit)? = null): T
    {
        val gson = createGson(visitor)
        return gson.fromJson(json, type)
    }

    @JvmOverloads
    @Throws(Exception::class)
    fun <T> fromJsonRes(
            context: Context,
            res: Int,
            type: Class<T>,
            visitor: ((builder: GsonBuilder) -> Unit)? = null
            ): T?
    {
        var obj: T?
        var bout: ByteArrayOutputStream? = null
        var fin: InputStream? = null

        try
        {
            fin = context.resources.openRawResource(res)
            bout = ByteArrayOutputStream()

            val buffer = ByteArray(256)
            var index = 0

            while (true)
            {
                index = fin.read(buffer)
                if (index < 0)
                {
                    break
                }

                bout.write(buffer, 0, index)
                bout.flush()
            }

            obj = fromJson(
                     String(bout.toByteArray()),
                     type,
                     visitor)
        }
        catch (ex: Exception)
        {
            throw ex
        }
        finally
        {
            try
            {
                fin?.close()
            }
            catch (ex: Exception)
            {
            }

            try
            {
                bout?.close()
            }
            catch (ex: Exception)
            {
            }
        }

        return obj
    }

    @JvmOverloads
    @Throws(Exception::class)
    fun <T> fromJsonRes(
            context: Context,
            res: Int,
            token: TypeToken<T>,
            visitor: ((builder: GsonBuilder) -> Unit)?= null
            ): T?
    {
        var obj: T?
        var bout: ByteArrayOutputStream? = null
        var fin: InputStream? = null

        try
        {
            fin = context.resources.openRawResource(res)
            bout = ByteArrayOutputStream()

            val buffer = ByteArray(256)
            var index = 0

            while (true)
            {
                index = fin.read(buffer)
                if (index < 0)
                {
                    break
                }

                bout.write(buffer, 0, index)
                bout.flush()
            }

            obj = fromJson(
                     String(bout.toByteArray()),
                     token,
                     visitor)
        }
        catch (ex: Exception)
        {
            throw ex
        }
        finally
        {
            try
            {
                fin?.close()
            }
            catch (ex: Exception)
            {
            }

            try
            {
                bout?.close()
            }
            catch (ex: Exception)
            {
            }
        }

        return obj
    }
}
